src/Symfony/Component/Security/Http/Firewall/UsernamePasswordJsonAuthenticationListener.php

```diff
@@ -0,0 +1,154 @@
+
+        try {
+            $username = $this->propertyAccessor->getValue($data, $this->options['username_path']);
+        } catch (AccessException $e) {
+            throw new BadCredentialsException(sprintf('Missing key "%s".', $this->options['username_path']));
+        }
+
+        try {
+            $password = $this->propertyAccessor->getValue($data, $this->options['password_path']);
+        } catch (AccessException $e) {
+            throw new BadCredentialsException(sprintf('Missing key "%s".', $this->options['password_path']));
+        }
+
+        if (!is_string($username)) {
+            throw new BadCredentialsException(sprintf('The key "%s" must contain a string.', $this->options['username_path']));
+        }
+
+        if (strlen($username) > Security::MAX_USERNAME_LENGTH) {
+            throw new BadCredentialsException('Invalid username.');
+        }
+
+        if (!is_string($password)) {
+            throw new BadCredentialsException(sprintf('The key "%s" must contain a string.', $this->options['password_path']));
+        }
+
+        try {
+            $token = new UsernamePasswordToken($username, $password, $this->providerKey);
+
+            $this->authenticationManager->authenticate($token);
+            $response = $this->onSuccess($request, $token);
+        } catch (AuthenticationException $e) {
+            $response = $this->onFailure($request, $e);
+        }
+
+        $event->setResponse($response);
+    }
+
+    private function onSuccess(Request $request, TokenInterface $token)
+    {
+        if (null !== $this->logger) {
+            $this->logger->info('User has been authenticated successfully.', array('username' => $token->getUsername()));
+        }
+
+        $this->tokenStorage->setToken($token);
+
+        if (null !== $this->eventDispatcher) {
+            $loginEvent = new InteractiveLoginEvent($request, $token);
+            $this->eventDispatcher->dispatch(SecurityEvents::INTERACTIVE_LOGIN, $loginEvent);
+        }
+
+        $response = $this->successHandler->onAuthenticationSuccess($request, $token);
+
+        if (!$response instanceof Response) {
+            throw new \RuntimeException('Authentication Success Handler did not return a Response.');
+        }
+
+        return $response;
+    }
+
+    private function onFailure(Request $request, AuthenticationException $failed)
+    {
+        if (null !== $this->logger) {
+            $this->logger->info('Authentication request failed.', array('exception' => $failed));
+        }
+
+        $token = $this->tokenStorage->getToken();
+        if ($token instanceof UsernamePasswordToken && $this->providerKey === $token->getProviderKey()) {
+            $this->tokenStorage->setToken(null);
+        }
+
+        $response = $this->failureHandler->onAuthenticationFailure($request, $failed);
+
+        if (!$response instanceof Response) {
+            throw new \RuntimeException('Authentication Failure Handler did not return a Response.');
```

Comment:
The point is to have the same behavior for this listener than for listeners extending the abstract one. Listeners not extending the abstract one are usually for special type of authentication. This is not the case here, it doesn't extends the abstract because of some implementation details (not the same set of dependencies...) but it shares the same goal.